db/system_distributed_keyspace.cc

```diff
@@ -92,6 +92,37 @@ schema_ptr cdc_generations() {
     return schema;
 }
 
+/* An internal table used by nodes to exchange CDC generation data. */
+schema_ptr cdc_generations_v2() {
+    thread_local auto schema = [] {
+        auto id = generate_legacy_id(system_distributed_keyspace::NAME_EVERYWHERE, system_distributed_keyspace::CDC_GENERATIONS_V2);
+        return schema_builder(system_distributed_keyspace::NAME_EVERYWHERE, system_distributed_keyspace::CDC_GENERATIONS_V2, {id})
+                /* The unique identifier of this generation. */
+                .with_column("id", uuid_type, column_kind::partition_key)
+                /* The generation describes a mapping from all tokens in the token ring to a set of stream IDs.
+                 * This mapping is built from a bunch of smaller mappings, each describing how tokens in a subrange
+                 * of the token ring are mapped to stream IDs; these subranges together cover the entire token ring.
+                 * Each such range-local mapping is represented by a row of this table.
+                 * The clustering key of the row is the end of the range being described by this row.
+                 * The start of this range is the range_end of the previous row (in the clustering order, which is the integer order)
+                 * or of the last row of this partition if this is the first the first row. */
+                .with_column("range_end", long_type, column_kind::clustering_key)
+                /* The set of streams mapped to in this range. */
+                .with_column("streams", cdc_streams_set_type)
+                /* The value of the `ignore_msb` sharding parameter of the node which was the owner of this token range
+                 * when the generation was first created. Together with the set of streams above it fully describes
+                 * the mapping for this particular range. */
+                .with_column("ignore_msb", byte_type)
+                /* Column used for sanity checking.
+                 * For a given generation it's equal to the number of ranges in this generation;
+                 * thus, after the generation is fully inserted, it must be equal to the number of rows in the partition. */
+                .with_column("num_ranges", int32_type, column_kind::static_column)
```

Comment:
Is this duplication necessary? The reader can get the same infomation by counting rows (or SELECT COUNT(...)). Having duplication in the data model can be okay, but has to be justified.

I see you say "sanity checking". Is it a way to "seal" the record?

Reply:
> I see you say "sanity checking". Is it a way to "seal" the record?

In a sense. I use it in `read_cdc_generation()`:
```
    // Paranoic sanity check. Partial reads should not happen since generations should be retrieved only after they
    // were written successfully with CL=ALL. But nobody uses EverywhereStrategy tables so they weren't ever properly
    // tested, so just in case...
    if (entries.size() != num_ranges) {
        throw std::runtime_error(format(
                "read_cdc_generation: wrong number of rows. The `num_ranges` column claimed {} rows,"
                " but reading the partition returned {}.", num_ranges, entries.size()));
    }
```
so this information is only used to increase confidence that the data is consistent. For example, suppose the creator of the data partially inserted the generation into the table due to a failure and then due to a bug started to gossip its ID (which it should not have because of the failure). Then we will detect that it's invalid and won't return it outside.

This actually already helped me catch one bug quickly where I accidentally performed a full-range-scan query in `read_cdc_generation` (so it retrieved all rows - of all generations) instead of selecting a particular generation; the sanity check catched the problem when there were 2 generations.

Previously the situation was simpler because we had one generation = one row.

Reply:
Ok. Too bad we don't have interactive transactions.

Reply:
What does happen if this table is partially written, then a crash happens?

Reply:
> What does happen if this table is partially written, then a crash happens?

It will contain a garbage partition that no node will ever try to retrieve, and if they do - due to some bug - they will most probably hit this sanity check.

The user is free to delete this garbage partition. We could potentially also introduce a background process that clears up such garbage generations (see e.g. #7300).